src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -98,88 +93,104 @@ public ImmutableArray<BoundDecisionDagNode> TopologicallySortedNodes
+                BoundDecisionDagNode makeReplacement(BoundDecisionDagNode dag, Func<BoundDecisionDagNode, BoundDecisionDagNode> replacement)
                 {
-                    case BoundEvaluationDecisionDagNode p:
-                        return p.Update(p.Evaluation, replacement[p.Next]);
-                    case BoundTestDecisionDagNode p:
+                    if (dag is BoundTestDecisionDagNode p)
+                    {
                         // This is the key to the optimization. The result of a top-level test might be known if the input is constant.
                         switch (knownResult(p.Test))
                         {
                             case true:
-                                return replacement[p.WhenTrue];
+                                return replacement(p.WhenTrue);
                             case false:
-                                return replacement[p.WhenFalse];
-                            default:
-                                return p.Update(p.Test, replacement[p.WhenTrue], replacement[p.WhenFalse]);
-                        }
-                    case BoundWhenDecisionDagNode p:
-                        if (p.WhenExpression == null || p.WhenExpression.ConstantValue == ConstantValue.True)
-                        {
-                            return p.Update(p.Bindings, p.WhenExpression, replacement[p.WhenTrue], null);
+                                return replacement(p.WhenFalse);
                         }
-                        else if (p.WhenExpression.ConstantValue == ConstantValue.False)
-                        {
-                            // It is possible in this case that we could eliminate some predecessor nodes, for example
-                            // those that compute evaluations only needed to get to this decision. We do not bother,
-                            // as that optimization would only be likely to affect test code.
-                            return replacement[p.WhenFalse];
-                        }
-                        else
-                        {
-                            return p.Update(p.Bindings, p.WhenExpression, replacement[p.WhenTrue], replacement[p.WhenFalse]);
-                        }
-                    case BoundLeafDecisionDagNode p:
-                        return p;
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(dag);
-                }
-            }
+                    }
 
-            // Is the decision's result known because the input is a constant?
-            bool? knownResult(BoundDagTest choice)
-            {
-                if (choice.Input.Source != null)
-                {
-                    // This is a test of something other than the main input; result unknown
-                    return null;
+                    return TrivialReplacement(dag, replacement);
```

Comment:
Yes, in the simplification call to Rewrite when we change the decision dag to avoid creating a tuple when switching on a tuple literal in a later iteration of this PR.

---
In reply to: [180571876](https://github.com/dotnet/roslyn/pull/25963#discussion_r180571876) [](ancestors = 180571876)